src/clj/puppetlabs/services/legacy_routes/legacy_routes_core.clj

```diff
@@ -0,0 +1,124 @@
+  not conform to RFC-2616 Accept: header specification.
+
+  The split on `,` is naive, the method should take into account `;` parameters
+  and such to be RFC compliant, but puppet agent does not use this format so
+  implementing the full spec is unnecessary."
+  [request]
+  (if-let [accept (get-in request [:headers "accept"])]
+    (let [vals (map str/lower-case (str/split accept #"\s*,\s*"))]
+      (assoc-in request [:headers "accept"]
+        (->> vals
+          (replace {"raw" "binary", "s" "binary"})
+          (distinct)
+          (str/join ", "))))
+    request))
+
+(defn- respond-with-content-type-text-plain
+  "Return a new ring handler with a ring middleware wrapped around `handler`
+  that unconditionally adds the Content-Type: text/plain header to the
+  response."
+  [handler]
+  (fn [request]
+    (ring-response/content-type (handler request) "text/plain")))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Handler wrapper
+
+(defn request-compatibility-wrapper
+  "Given a Puppet 3.x request, convert it to a Puppet 4.x request."
+  [handler mount-point api-version request]
+  (let [{{environment :environment} :params
+         path-info                  :path-info
+         query-string               :query-string} request
+        query-params (if (str/blank? query-string)
+                       {}
+                       (ring-codec/form-decode query-string))]
+    (let [compat-request
+          (-> request
+              (map-accept-header)
+              (assoc :path-info    (str "/" api-version path-info)
+                     :context      mount-point
+                     :uri          (str mount-point "/" api-version path-info)
+                     :query-string (ring-codec/form-encode
+                                     (merge
+                                       query-params
+                                       (when environment {:environment environment}))))
+              (dissoc :params :route-params))]
+      (handler compat-request))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Routing
+
+(defn v2_0-routes
+  "Creates the compojure routes to handle the 3.X master's '/v2.0' routes."
+  [request-handler]
+  (compojure/routes
+    (compojure/GET "/environments" request
+                   (request-handler request))))
+
+(defn legacy-master-routes
+  [master-request-handler ca-request-handler]
+  (compojure/routes
+    (compojure/GET "/node/*" request (master-request-handler request))
+    (compojure/GET "/file_content/*" request (master-request-handler request))
+    (compojure/GET "/file_metadatas/*" request (master-request-handler request))
+    (compojure/GET "/file_metadata/*" request (master-request-handler request))
+    (compojure/GET "/file_bucket_file/*" request
+      ((respond-with-content-type-text-plain master-request-handler) request))
```

Comment:
@nwolfe @camlow325 This should fix the `puppet filebucket restore ...` issue you found.  :pray: